oci/layer/tar_extract.go

```diff
@@ -372,12 +387,22 @@ func (te *TarExtractor) overlayFSWhiteout(dir string, file string) error {
 	}
 
 	// otherwise, white out the file itself.
+
 	p := filepath.Join(dir, strings.TrimPrefix(file, whPrefix))
-	if err := os.RemoveAll(p); err != nil && !os.IsNotExist(err) {
-		return errors.Wrapf(err, "couldn't create overlayfs whiteout for %s", p)
-	}
 
-	err := te.fsEval.Mknod(p, unix.S_IFCHR|0666, unix.Mkdev(0, 0))
+	var err error
+
+	if ok, err := isEmptyDir(p); ok && err == nil {
+		// dir is empty, then remove and create the overlay char dev
+		if err := os.RemoveAll(p); err != nil && !os.IsNotExist(err) {
+			return errors.Wrapf(err, "couldn't create overlayfs whiteout for %s", p)
+		}
+
+		err = te.fsEval.Mknod(p, unix.S_IFCHR|0666, unix.Mkdev(0, 0))
+	} else {
+		// subdirs already exist, then no need to remove, just add the opaque attr
+		err = te.fsEval.Lsetxattr(p, "user.overlay.opaque", []byte("y"), 0)
```

Comment:
I think we might need to add a check that `p` is actually a directory.
If `p` is a regular file and not a directory I think we end up on this branch (because `isEmptyDir` will return an error from Readdirnames I assume) and just set an attr on the file, which isn't what we want for regular files, right?